segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLog.java

```diff
@@ -635,18 +655,27 @@ private LogMetadata loadMetadata() throws DataLogInitializationException {
     /**
      * Updates the metadata and persists it as a result of adding a new Ledger.
      *
-     * @param currentMetadata The current metadata.
-     * @param newLedger       The newly added Ledger.
+     * @param currentMetadata   The current metadata.
+     * @param newLedger         The newly added Ledger.
+     * @param clearEmptyLedgers If true, the new metadata will not not contain any pointers to empty Ledgers. Setting this
+     *                          to true will not remove a pointer to the last few ledgers in the Log (controlled by
+     *                          Ledgers.MIN_FENCE_LEDGER_COUNT, even if they are indeed empty (this is so we don't interfere
+     *                          with any ongoing fencing activities as another instance of this Log may not have yet been
+     *                          fenced out).
      * @return A new instance of the LogMetadata, which includes the new ledger.
      * @throws DurableDataLogException If an Exception occurred.
      */
-    private LogMetadata updateMetadata(LogMetadata currentMetadata, LedgerHandle newLedger) throws DurableDataLogException {
+    private LogMetadata updateMetadata(LogMetadata currentMetadata, LedgerHandle newLedger, boolean clearEmptyLedgers) throws DurableDataLogException {
```

Comment:
I think that we need this flag because we only want to clean up empty ledgers upon initialization, while metadata updates can occur at other times too. Is my understanding correct?

Reply:
Yes. We update metadata upon truncation too. I don't want to do this "cleanup" then because it would be useless anyway. I only want to do this upon initialization, because that's when we are most likely to run into empty ledgers due to incomplete recoveries.

Reply:
ok.